packages: skip loading already-loaded kernel modules

msvamp opened the pull request:
Created a new function that loads kernel modules only if not already loaded
This prevents lines like -
   daemon.err insmod: module is already loaded - slhc
from appearing the device logs everytime a PPP connection reconnects

Signed-off-by: Manas Sambhus <51392775+msvamp@users.noreply.github.com>

obsy commented:
Nice idea. But maybe it should go to /lib/functions.sh as generic function, and you should change also other scripts like 15-teql in   iproute2, generate.sh in qos-scripts etc?

msvamp commented:
Thanks for the tip @obsy!

I'll take some time and look for all occurrences of `insmod` in various packages and make them use a function from the generic functions script.

adschm commented:
Please bump PKG_RELEASE.

msvamp commented:
I was wondering if there is an advantage using `modprobe` rather than `insmod`?
It does not produce any error message at least, when an existing module is loaded.

CodeFetch commented:
Ordinarily modprobe is less error-prone as it automatically loads dependencies for missing symbols.

adschm commented:
As I understand it, *noreply* e-mail addresses don't work for sending mails to them, so please replace with something that works.

msvamp commented:
> Ordinarily modprobe is less error-prone as it automatically loads dependencies for missing symbols.

@CodeFetch Thanks, is there any performance implication that one should worry about while using `modprobe`?



> As I understand it, _noreply_ e-mail addresses don't work for sending mails to them, so please replace with something that works.

@adschm I didn't quite understand whom are you referring to by "them"

adschm commented:
> @adschm I didn't quite understand whom are you referring to by "them"

The noreply e-mail addresses ...

The point of the mail address in Signed-off-by is to reach the person. Will I reach you when I send an e-mail to 51392775+msvamp@users.noreply.github.com ?

CodeFetch commented:
@msvamp No, but you should be aware that it loads the dependencies automatically. So possibly you don't need to specify ppp_generic I guess.

guidosarducci commented:
Having had to fix problems there in the past, I really think modifying the core shell library code without strong reason, clear benefits, and a lot of testing should be avoided. Adding a library function seems too heavy-weight to simply clean up a log message, when you could just update a package to do `modprobe -q <module>`.

Also, is the `iproute2` change related to an `insmod` log error? I can't recall seeing that before so wondering if necessary... Thanks!

CodeFetch commented:
@guidosarducci Sarcasm? I just think insmod is not a good choice from the beginning. 
Man page of insmod:
```
       insmod is a trivial program to insert a module into the kernel.
       Most users will want to use modprobe(8) instead, which is more
       clever and can handle module dependencies.
```

guidosarducci commented:
> Having had to fix problems there in the past, I really think modifying the core shell library code without strong reason, clear benefits, and a lot of testing should be avoided. Adding a library function seems too heavy-weight to simply clean up a log message, when you could just update a package to do `modprobe -q <module>`.
> 
> Also, is the `iproute2` change related to an `insmod` log error? I can't recall seeing that before so wondering if necessary... Thanks!

@CodeFetch This is clearly a comment and question on the PR, for the benefit of the author, and without any discernible sarcasm. It doesn't mention you or even suggest using `insmod`, so your response and man-page snippet don't really follow at all. If you're simply trolling or seeking an argument, then I'm not interested.

CodeFetch commented:
@guidosarducci 
No, I really thought this was sarcasm

> I really think modifying the core shell library code without strong reason, clear benefits, and a lot of testing should be avoided.

Points on sarcasm in this sentence:
- modifying the core shell library code (understood: sounds like something that could break a lot!)
- without strong reason (understood: better think about it!)
- clear benefits (understood: where is the benefit?)
- a lot of testing should be avoided (understood: test it thoroughly!)

msvamp commented:
> The point of the mail address in Signed-off-by is to reach the person.

@adschm I believe this patch is quite trivial for anyone to contact me. Also, I am [not alone](https://github.com/openwrt/openwrt/search?q=noreply&type=commits) in using a `noreply` address. But I agree wth you, I'll replace my commits with a reachable address or send patches to the mailing lists instead.

---

@CodeFetch and @guidosarducci Thank you for your suggestions.

> Adding a library function seems too heavy-weight

Heavier than replacing `insmod` with `modprobe` (which also does things like dependency resolution) ?

I am completely a newbie here, so I'm not sure of the importance of the _core library code_ and what kind of functions should go in, and what not. The suggestion of adding it there was made by @obsy and I assumed it should probably be the right place because I did see some trivial functions here, which could have _just been updated in the respective packages_.  
For example: [`debug`](https://github.com/openwrt/openwrt/blob/ce4cb8e51d8d4b24267653250c3f7310d208bfaa/package/base-files/files/lib/functions.sh#L6), [`config_unset`](https://github.com/openwrt/openwrt/blob/ce4cb8e51d8d4b24267653250c3f7310d208bfaa/package/base-files/files/lib/functions.sh#L102), [`config_set`](https://github.com/openwrt/openwrt/blob/ce4cb8e51d8d4b24267653250c3f7310d208bfaa/package/base-files/files/lib/functions.sh#L139), [`group_exists`](https://github.com/openwrt/openwrt/blob/ce4cb8e51d8d4b24267653250c3f7310d208bfaa/package/base-files/files/lib/functions.sh#L326), [`user_exists`](https://github.com/openwrt/openwrt/blob/ce4cb8e51d8d4b24267653250c3f7310d208bfaa/package/base-files/files/lib/functions.sh#L379)

But anyway, I get your point. This file is going to be included in several other places where `insmod` has no use, and hence, defining the function is simply wasting resources.

> Also, is the iproute2 change related to an insmod log error? I can't recall seeing that before so wondering if necessary

I am not really sure as I have never seen that error in my log either. I just assumed someone out there might have seen it (because it too, loads the modules unchecked) and this should also fix it for them.

adschm commented:
Ping.

msvamp commented:
Oops, I updated my fork and it accidentally closed this PR. I will make the new changes and reopen this PR at the earliest.

P.S. The old set of changes done using an `insmod_ignore` function have been migrated to [another branch on my fork](https://github.com/msvamp/openwrt/tree/packages-fix-insmod).

adschm commented:
https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr

msvamp committed:
ppp: use modprobe in place of insmod

This will prevent `module is already loaded` lines from
appearing in the logs when a PPP connection is reconnecting

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

msvamp committed:
kernel: netsupport: replace insmod by modprobe

Replace insmod by modprobe in TEQL hotplug script

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

msvamp committed:
qos-scripts: replace modprobe by rmmod

modprobe -r is not available on all platforms, hence use rmmod

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

Ansuel commented:
@msvamp considering the problem with modprobe or other utility not present... can we use this as an alternative?
https://stackoverflow.com/questions/9845877/how-to-determine-if-a-specific-module-is-loaded-in-linux-kernel
Seems more safe that mod the bin to use :D 

Also i notice in one commit you fix whitespace problem while tweaking the insmod... Can you split that commit and fix the spacing problem in a different commit? (just to keep things commits specific to the single task)

msvamp commented:
Hey @Ansuel,

> can we use this as an alternative?
https://stackoverflow.com/questions/9845877/how-to-determine-if-a-specific-module-is-loaded-in-linux-kernel

I had implemented something similar by adding a core library function [`insmod_ignore`](https://github.com/msvamp/openwrt/commit/71909ab1) in my previous set of commits about an year ago, however I discovered that some packages are already using `modprobe` and further @guidosarducci also suggested that it's a better option.

Are there really any platforms where `modprobe` is not available? I have seen that certain options of modprobe, like `-r` might be missing on certain platforms (like ramips for example) but the basic module loading functionality is present.

> Seems more safe that mod the bin to use

_mod the bin_? Sorry, I didn't get you.

> Can you split that commit and fix the spacing problem in a different commit?

Sure, once the above approach is decided, I'll separate the whitespace part in a separate commit.

Ansuel commented:
@msvamp ok lets use modprobe then

msvamp committed:
qos-scripts: fix trailing whitespace in config files

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

msvamp commented:
@Ansuel I have separated the whitespace fixes into a separate commit and also performed it on other files in the package.

Ansuel commented on package/network/config/qos-scripts/files/usr/lib/qos/generate.sh:
```diff
@@ -2,11 +2,10 @@
 	insmod="insmod"
-	rmmod="rmmod"
 }
+rmmod="rmmod"
```
Hi, hope you don't mind... Considering we are dropping this, can't we just hardcode and drop the variable?

Ansuel commented on package/network/services/ppp/Makefile:
```diff
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=ppp
-PKG_RELEASE:=3
+PKG_RELEASE:=4
```
Can you change this to autorelease api (if you don't know what i'm talking about feel free to ask) in another commit? (ideally before this?)

Ansuel replied:
(example https://github.com/openwrt/openwrt/blob/master/package/network/config/qos-scripts/Makefile#L12)

Ansuel commented:
Merged in master with
3e2ea10e5e qos-scripts: fix trailing whitespace in config files
0ca634e9ef qos-scripts: replace modprobe by rmmod
4cc7011da0 kernel: netsupport: replace insmod by modprobe
db0c0a31d8 ppp: use modprobe in place of insmod

Thanks!

msvamp commented:
Sorry, I was busy with some work, and didn't get a chance to check. What changes were you asking for? I'll try to implement them in the next month, when I'll get some time.

Also, thanks for merging
